lib/Targets/AIETargetAirbin.cpp

```diff
@@ -0,0 +1,1746 @@
+static void removeWritesInRange(Address start, Address end) {
+  auto iter =
+      std::remove_if(writes.begin(), writes.end(), [&](const Write &write) {
+        return start <= write.destination() and write.destination() <= end;
+      });
+
+  if (iter != writes.end())
+    writes.erase(iter, writes.end());
+}
+
+// Set every address in the (inclusive on both ends) range to 0.
+// This function is faster and prefered over the equivalent `for` over
+// `write32`, due to batching.
+static void clearRange(TileAddress tile, uint32_t range_start,
+                       uint32_t range_end) {
+  assert(range_start <= range_end);
+  assert(range_start % 4 == 0);
+  assert(range_end % 4 == 0);
+
+  // Since we are emplacing a chunk of contiguous writes,
+  // it is more efficient to
+  // 1. reserve extra memory
+  // 2. remove all to-be-overwritten writes ahead of time
+  // 3. directly add the writes
+  // than to use `write32` repeatedly.
+  reserveWrites((range_end - range_start) / 4);
+  removeWritesInRange({tile, range_start}, {tile, range_end});
+
+  for (auto off = range_start; off <= range_end; off += 4u) {
+    writes.emplace_back(Address{tile, off}, 0);
+  }
+}
+
+static constexpr auto BD_BASE = 0x1D000u;
+static constexpr auto BD_END = 0x1D1F8u;
+static constexpr auto SHIM_BD_END = 0x1D15Cu;
+
+// The SHIM row is always 0.
+// SHIM resets are handled by the runtime.
+static void generateShimConfig(TileOp &tileOp) {
+
+  TileAddress tileAddress{tileOp};
+
+  if (tileOp.isShimNOCTile()) {
+    clearRange(tileAddress, BD_BASE, SHIM_BD_END);
+  }
+  if (tileOp.isShimNOCTile() or tileOp.isShimTile()) {
+    // output << "// Stream Switch master config\n";
+    clearRange(tileAddress, 0x3F000, 0x3F058);
+    // output << "// Stream Switch slave config\n";
+    clearRange(tileAddress, 0x3F100, 0x3F15C);
+    // output << "// Stream Switch slave slot config\n";
+    clearRange(tileAddress, 0x3F200, 0x3F368);
+  }
+}
+
+// This template can be instantiated to represent a bitfield in a register.
+template <uint8_t high_bit, uint8_t low_bit = high_bit> class Field final {
+public:
+  static_assert(high_bit >= low_bit,
+                "The high bit should be higher than the low bit");
+  static_assert(high_bit < sizeof(uint32_t) * 8u,
+                "The field must live in a 32-bit register");
+
+  static constexpr auto num_bits_used = (high_bit - low_bit) + 1u;
+  static constexpr auto unshifted_mask = (1u << num_bits_used) - 1u;
+  static_assert((low_bit != high_bit) xor (unshifted_mask == 1),
+                "1 is a valid mask iff the field is 1 bit wide");
+
+  static constexpr auto shifted_mask = unshifted_mask << low_bit;
+
+  [[nodiscard]] constexpr uint32_t operator()(uint32_t value) const {
+    return (value << low_bit) & shifted_mask;
+  }
+};
+
+// LLVM may have a way to read ELF files,
+// but for our purposes, manually parsing is fine for now.
+static bool loadElf(TileAddress tile, const std::string &filename) {
```

Comment:
Um... yeah... There's enough magic stuff here to use the LLVM way... you want:

  Expected<std::unique_ptr<ObjectFile>> ELFObjOrErr =
      object::ObjectFile::createELFObjectFile(
          MemoryBufferRef(toStringRef(D.Data), "dummyELF"));

See llvm/object/ObjectFile.h and llvm/object/ELF.h